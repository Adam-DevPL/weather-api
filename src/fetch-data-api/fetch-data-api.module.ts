import { HttpModule } from '@nestjs/axios';
import { Module } from '@nestjs/common';
import { FetchDataApiFilter } from 'src/filters/fetch-data-api.filter';
import { FetchDataApiService } from './fetch-data-api.service';
import { APP_FILTER } from '@nestjs/core';

@Module({
  imports: [HttpModule],
  providers: [
    FetchDataApiService,
    {
      provide: APP_FILTER,
      useClass: FetchDataApiFilter,
    },
  ],
  exports: [FetchDataApiService],
})
export class FetchDataApiModule {}
