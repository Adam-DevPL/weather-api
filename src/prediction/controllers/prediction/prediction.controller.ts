import {
  Controller,
  Get,
  HttpException,
  HttpStatus,
  Param,
} from '@nestjs/common';
import { DateParam } from 'src/decorators/DataDecorator';
import { PredictionService } from 'src/prediction/services/prediction/prediction.service';
import {
  PredictionForecastResponse,
  PredictionRouteCityDayParams,
  PredictionRouteCountryDayParams,
} from 'src/prediction/types/prediction.types';

@Controller('weather/prediction')
export class PredictionController {
  constructor(private predictionService: PredictionService) {}

  @Get('country/:day/:country')
  getForecastForCountry(
    @DateParam('day') day: Date,
    @Param('country') country: string,
  ): Promise<PredictionForecastResponse> {
    if (!day) {
      throw new HttpException(`Invalid date`, HttpStatus.BAD_REQUEST);
    }
    const predictionRouteCountryDayParams: PredictionRouteCountryDayParams = {
      day: day.toISOString().split('T')[0],
      country,
    };
    return this.predictionService.getForecastForCountry(
      predictionRouteCountryDayParams,
    );
  }
  @Get('city/:day/:city')
  getForecastForCity(
    @DateParam('day') day: Date,
    @Param('city') city: string,
  ): Promise<PredictionForecastResponse> {
    if (!day) {
      throw new HttpException(`Invalid date`, HttpStatus.BAD_REQUEST);
    }
    const predictionRouteCityDayParams: PredictionRouteCityDayParams = {
      day: day.toISOString().split('T')[0],
      city,
    };

    return this.predictionService.getForecastForCity(
      predictionRouteCityDayParams,
    );
  }
}
