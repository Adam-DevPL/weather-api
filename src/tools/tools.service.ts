import { Injectable } from '@nestjs/common';

@Injectable()
export class ToolsService {
  public getAverage = (arr: number[]): number => {
    return arr.reduce((acc, item) => acc + item, 0) / arr.length;
  };

  public mostFrequent = (arr: number[]): string =>
    Object.entries(
      arr.reduce((a, v) => {
        a[v] = a[v] ? a[v] + 1 : 1;
        return a;
      }, {}),
    ).reduce((a, v) => (v[1] >= a[1] ? v : a), [null, 0])[0];

  public getWeatherCode = (code: number): string => {
    switch (code) {
      case 0:
        return 'Clear sky';
      case 1:
      case 2:
      case 3:
        return 'Mainly clear, partly cloudy, and overcast';
      case 45:
      case 48:
        return 'Fog and depositing rime fog';
      case 51:
      case 53:
      case 55:
        return 'Drizzle: Light, moderate, and dense intensity';
      case 56:
      case 57:
        return 'Freezing Drizzle: Light and dense intensity';
      case 61:
      case 63:
      case 65:
        return 'Rain: Slight, moderate and heavy intensity';
      case 66:
      case 67:
        return 'Freezing Rain: Light and heavy intensity';
      case 71:
      case 73:
      case 75:
        return 'Snow fall: Slight, moderate, and heavy intensity';
      case 77:
        return 'Snow grains';
      case 80:
      case 81:
      case 82:
        return 'Rain showers: Slight, moderate, and violent';
      case 85:
      case 86:
        return 'Snow showers slight and heavy';
      case 95:
        return 'Thunderstorm: Slight or moderate';
      case 96:
      case 99:
        return 'Thunderstorm with slight and heavy hail';
      default:
        return 'No more weather details';
    }
  };
}
