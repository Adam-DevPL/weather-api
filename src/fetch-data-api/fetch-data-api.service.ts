import { HttpService } from '@nestjs/axios';
import { HttpException, HttpStatus, Injectable } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';

import { firstValueFrom } from 'rxjs';
import { LocationType } from 'src/prediction/types/prediction.types';
import {
  FetchDataApiGeoResponse,
  FetchDataApiParams,
} from './types/FetchDataApi.types';

@Injectable()
export class FetchDataApiService {
  private readonly weatherUrl;
  private readonly geoUrl;

  constructor(
    private readonly httpService: HttpService,
    private readonly configService: ConfigService,
  ) {
    this.weatherUrl = this.configService.get<string>('WEATHER_URL');
    this.geoUrl = this.configService.get<string>('GEO_URL');
  }

  public async getDataFromApi(params: FetchDataApiParams) {
    try {
      const { data } = await firstValueFrom(
        this.httpService.get(this.weatherUrl, { params }),
      );
      if (data.error) {
        throw new HttpException(
          { status: HttpStatus.BAD_REQUEST, message: 'Invalid coordinates' },
          HttpStatus.BAD_REQUEST,
        );
      }

      return {
        data,
      };
    } catch ({ response }) {
      throw new HttpException(
        {
          status:
            response.status === 404
              ? HttpStatus.INTERNAL_SERVER_ERROR
              : HttpStatus.BAD_REQUEST,
          error:
            response.status === 404
              ? 'Internal server error'
              : response.message,
        },
        response.status === 404
          ? HttpStatus.INTERNAL_SERVER_ERROR
          : HttpStatus.BAD_REQUEST,
        {
          cause: response,
        },
      );
    }
  }

  public async getGeoLocation(name: string): Promise<FetchDataApiGeoResponse> {
    try {
      const params = {
        name,
        count: 1,
      };

      const { data } = await firstValueFrom(
        this.httpService.get(this.geoUrl, { params }),
      );

      if (!data.hasOwnProperty('results')) {
        throw new HttpException(
          { status: HttpStatus.BAD_REQUEST, message: 'Location not found' },
          HttpStatus.BAD_REQUEST,
        );
      }

      const {
        results: [{ latitude, longitude, country }],
      } = data;

      return {
        latitude,
        longitude,
        location:
          country.toLowerCase() === name.toLowerCase()
            ? LocationType.Country
            : LocationType.City,
      };
    } catch ({ response }) {
      throw new HttpException(
        {
          status:
            response.status === 404
              ? HttpStatus.INTERNAL_SERVER_ERROR
              : HttpStatus.BAD_REQUEST,
          error:
            response.status === 404
              ? 'Internal server error'
              : response.message,
        },
        response.status === 404
          ? HttpStatus.INTERNAL_SERVER_ERROR
          : HttpStatus.BAD_REQUEST,
        {
          cause: response,
        },
      );
    }
  }
}
