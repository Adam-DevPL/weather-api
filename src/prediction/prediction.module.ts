import { Module } from '@nestjs/common';
import { FetchDataApiModule } from 'src/utils/fetch-data-api/fetch-data-api.module';
import { ToolsModule } from 'src/utils/tools/tools.module';
import { PredictionController } from './controllers/prediction/prediction.controller';
import { PredictionService } from './services/prediction/prediction.service';

@Module({
  imports: [FetchDataApiModule, ToolsModule],
  controllers: [PredictionController],
  providers: [PredictionService],
})
export class PredictionModule {}
