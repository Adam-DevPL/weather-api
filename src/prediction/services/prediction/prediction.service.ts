import { HttpException, HttpStatus, Injectable } from '@nestjs/common';
import {
  LocationType,
  OpenModelResponse,
  PredictionDataObject,
  PredictionForecastResponse,
  PredictionRouteLocationDateRangeParams,
  PredictionRouteLocationSingleDayParam,
  WeatherInfo,
} from 'src/prediction/types/prediction.types';
import { FetchDataApiService } from 'src/utils/fetch-data-api/fetch-data-api.service';
import { FetchDataApiParams } from 'src/utils/fetch-data-api/types/FetchDataApi.types';
import { ToolsService } from 'src/utils/tools/tools.service';

@Injectable()
export class PredictionService {
  constructor(
    private fetchDataApi: FetchDataApiService,
    private tools: ToolsService,
  ) {}

  public async getForecastForSingleDay({
    day,
    locationName,
    locationType,
  }: PredictionRouteLocationSingleDayParam): Promise<PredictionForecastResponse> {
    const { latitude, longitude, location } =
      await this.fetchDataApi.getGeoLocation(locationName);

    if (locationType !== location) {
      throw new HttpException(
        `It is not a ${LocationType[locationType]}: ${locationName}`,
        HttpStatus.BAD_REQUEST,
      );
    }

    const fetchDataApiParams: FetchDataApiParams = {
      latitude,
      longitude,
      start_date: day,
      end_date: day,
      hourly: 'temperature_2m,weathercode',
    };

    const {
      data: {
        hourly: { temperature_2m, weathercode },
      },
    } = await this.fetchDataApi.getDataFromApi(fetchDataApiParams);

    const avgTemperature: string = (
      temperature_2m.reduce((a: number, b: number) => a + b, 0) /
      temperature_2m.length
    ).toFixed(1);

    const weatherCode = this.tools.getWeatherCode(
      Number(this.tools.mostFrequent(weathercode)),
    );

    return {
      location: locationName,
      weatherInfo: { day: day, avgTemperature, weather: weatherCode },
    };
  }

  public async getForecastForLocationInDateRange({
    locationType,
    locationParam,
    from,
    to,
  }: PredictionRouteLocationDateRangeParams) {
    let lat: number;
    let lon: number;

    if (locationType !== LocationType.Geo) {
      const { latitude, longitude, location } =
        await this.fetchDataApi.getGeoLocation(locationParam);

      if (locationType !== location) {
        throw new HttpException(
          `It is not a ${LocationType[locationType]}: ${locationParam}`,
          HttpStatus.BAD_REQUEST,
        );
      }
      lat = latitude;
      lon = longitude;
    } else {
      lat = locationParam.lat;
      lon = locationParam.lon;
    }

    const fetchDataApiParams: FetchDataApiParams = {
      latitude: lat,
      longitude: lon,
      start_date: from,
      end_date: to,
      hourly: 'temperature_2m,weathercode',
    };

    const {
      data: {
        hourly: { time, temperature_2m, weathercode },
      },
    }: OpenModelResponse = await this.fetchDataApi.getDataFromApi(
      fetchDataApiParams,
    );

    const outputArr = time.map((element, index) => {
      const tmp: PredictionDataObject = {
        time: element.slice(0, element.indexOf('T')),
        temperature: temperature_2m[index],
        weatherCode: weathercode[index],
      };
      return tmp;
    });

    const grouped: Array<WeatherInfo> = Object.values(
      outputArr.reduce((acc, item) => {
        acc[item.time] = [...(acc[item.time] || []), item];
        return acc;
      }, {}),
    ).map((element: PredictionDataObject[]) => {
      const temperature = element.map((e) => e.temperature);
      const weatherCode = element.map((e) => e.weatherCode);
      const tmp: WeatherInfo = {
        day: element[0].time,
        avgTemperature: this.tools.getAverage(temperature).toFixed(1),
        weather: this.tools.getWeatherCode(
          Number(this.tools.mostFrequent(weatherCode)),
        ),
      };
      return tmp;
    });

    return {
      location: locationParam,
      weatherInfo: grouped,
    };
  }
}
