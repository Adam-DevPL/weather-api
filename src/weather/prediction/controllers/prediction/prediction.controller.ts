import {
  Controller,
  Get,
  HttpException,
  HttpStatus,
  Param,
} from '@nestjs/common';
import { PredictionService } from 'src/prediction/services/prediction/prediction.service';
import { ForecastResponse } from 'src/prediction/types/prediction.types';
import { ToolsService } from 'src/tools/tools.service';
import {
  CoordinatesInputDates,
  LocationNameInputDates,
  LocationNameInputDay,
  LocationType,
} from 'src/weather/types/weather.types';
import {
  CityDatesRangeParam,
  CityDayParam,
  CountryDatesRangeParam,
  CountryDayParam,
  GeoCoordinatesDatesRangeParam,
} from 'src/Validation/validation.module';

@Controller('weather/prediction')
export class PredictionController {
  constructor(
    private predictionService: PredictionService,
    private toolsService: ToolsService,
  ) {}

  @Get('country/:day/:country')
  getForecastForCountry(
    @Param() { country, day }: CountryDayParam,
  ): Promise<ForecastResponse> {
    const locationNameInputDay: LocationNameInputDay = {
      day,
      locationName: country,
      locationType: LocationType.COUNTRY,
    };

    return this.predictionService.getForecastForSingleDay(locationNameInputDay);
  }
  @Get('city/:day/:city')
  getForecastForCity(
    @Param() { city, day }: CityDayParam,
  ): Promise<ForecastResponse> {
    const locationNameInputDay: LocationNameInputDay = {
      day,
      locationName: city,
      locationType: LocationType.CITY,
    };

    return this.predictionService.getForecastForSingleDay(locationNameInputDay);
  }

  @Get('country/:country/:from/:to')
  getForecastForCountryInDateRange(
    @Param() { country, from, to }: CountryDatesRangeParam,
  ): Promise<ForecastResponse> {
    if (this.toolsService.checkDatesRange(from, to)) {
      throw new HttpException(
        'Invalid dates order. Date "from" is later then "to" date',
        HttpStatus.BAD_REQUEST,
      );
    }
    const locationNameInputDates: LocationNameInputDates = {
      locationType: LocationType.COUNTRY,
      locationName: country,
      from,
      to,
    };

    return this.predictionService.getForecastForCountryOrCityInDateRange(
      locationNameInputDates,
    );
  }

  @Get('city/:city/:from/:to')
  getForecastForCityInDateRange(
    @Param() { city, from, to }: CityDatesRangeParam,
  ): Promise<ForecastResponse> {
    if (this.toolsService.checkDatesRange(from, to)) {
      throw new HttpException(
        'Invalid dates order. Date "from" is later then "to" date',
        HttpStatus.BAD_REQUEST,
      );
    }

    const locationNameInputDates: LocationNameInputDates = {
      locationType: LocationType.CITY,
      locationName: city,
      from,
      to,
    };

    return this.predictionService.getForecastForCountryOrCityInDateRange(
      locationNameInputDates,
    );
  }

  @Get('location/:lat/:lon/:from/:to')
  getForecastForCoordinatesInDateRange(
    @Param() { lat, lon, from, to }: GeoCoordinatesDatesRangeParam,
  ): Promise<ForecastResponse> {
    if (this.toolsService.checkDatesRange(from, to)) {
      throw new HttpException(
        'Invalid dates order. Date "from" is later then "to" date',
        HttpStatus.BAD_REQUEST,
      );
    }
    const coordinatesInputDates: CoordinatesInputDates = {
      lat,
      lon,
      from,
      to,
    };

    return this.predictionService.getForecastForGeoLocationInDateRange(
      coordinatesInputDates,
    );
  }
}
