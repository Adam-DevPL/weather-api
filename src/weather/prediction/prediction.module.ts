import { Module } from '@nestjs/common';
import { FetchDataApiModule } from 'src/fetch-data-api/fetch-data-api.module';
import { ToolsModule } from 'src/tools/tools.module';
import { PredictionController } from './controllers/prediction.controller';
import { PredictionService } from './services/prediction.service';

@Module({
  imports: [FetchDataApiModule, ToolsModule],
  controllers: [PredictionController],
  providers: [PredictionService],
})
export class PredictionModule {}
