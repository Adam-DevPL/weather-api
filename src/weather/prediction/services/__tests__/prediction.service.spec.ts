import { Test, TestingModule } from '@nestjs/testing';
import { PredictionService } from '../prediction.service';
import { ConfigService } from '@nestjs/config';
import { HttpModule } from '@nestjs/axios';
import { FetchDataApiService } from 'src/fetch-data-api/fetch-data-api.service';
import { ToolsService } from 'src/tools/tools.service';
import {
  CoordinatesInputDates,
  LocationNameInputDates,
  LocationNameInputDay,
  LocationType,
} from 'src/weather/types/weather.types';
import { temperature, time, weathercode } from './testData';
import {
  InternalServerException,
  LocationException,
} from 'src/filters/exceptions/exceptions';

describe('PredictionService', () => {
  let service: PredictionService;
  let fetchDataApi: FetchDataApiService;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      imports: [HttpModule],
      providers: [
        PredictionService,
        FetchDataApiService,
        ToolsService,
        ConfigService,
      ],
    }).compile();

    service = module.get<PredictionService>(PredictionService);
    fetchDataApi = module.get<FetchDataApiService>(FetchDataApiService);
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
  });

  describe('getForecastForSingleDay', () => {
    it('should return weather response for valid country - forecast on 1 day', async () => {
      //given
      const locationNameInputDay: LocationNameInputDay = {
        day: '2023-02-15',
        locationName: 'country',
        locationType: LocationType.COUNTRY,
      };

      jest.spyOn(fetchDataApi, 'getGeoLocation').mockResolvedValue({
        latitude: 20,
        longitude: 20,
        location: LocationType.COUNTRY,
      });

      jest.spyOn(service, <any>'getDataApi').mockResolvedValue({
        temperatureArr: [10, 20, 30],
        weatherCodeArr: [2, 2, 2],
      });

      //when
      const result = await service.getForecastForSingleDay(
        locationNameInputDay,
      );

      //then
      expect(result).toMatchSnapshot();
    });

    it('should return weather response for valid city - forecast on 1 day', async () => {
      //given
      const locationNameInputDay: LocationNameInputDay = {
        day: '2023-02-15',
        locationName: 'city',
        locationType: LocationType.CITY,
      };

      jest.spyOn(fetchDataApi, 'getGeoLocation').mockResolvedValue({
        latitude: 20,
        longitude: 20,
        location: LocationType.CITY,
      });

      jest.spyOn(service, <any>'getDataApi').mockResolvedValue({
        temperatureArr: [10, 20, 30],
        weatherCodeArr: [2, 2, 2],
      });

      //when
      const result = await service.getForecastForSingleDay(
        locationNameInputDay,
      );

      //then
      expect(result).toMatchSnapshot();
    });

    it('should throw error, when wants country, but get city - forecast on 1 day', async () => {
      //given
      const locationNameInputDay: LocationNameInputDay = {
        day: '2023-02-15',
        locationName: 'city',
        locationType: LocationType.COUNTRY,
      };

      jest.spyOn(fetchDataApi, 'getGeoLocation').mockResolvedValue({
        latitude: 20,
        longitude: 20,
        location: LocationType.CITY,
      });

      await expect(
        service.getForecastForSingleDay(locationNameInputDay),
      ).rejects.toMatchSnapshot();
    });

    it('should throw error, when unnamed city - forecast on 1 day', async () => {
      //given
      const locationNameInputDay: LocationNameInputDay = {
        day: '2023-02-15',
        locationName: 'xxxxx',
        locationType: LocationType.CITY,
      };

      jest.spyOn(fetchDataApi, 'getGeoLocation').mockImplementation(() => {
        throw new LocationException('Location not found');
      });

      await expect(
        service.getForecastForSingleDay(locationNameInputDay),
      ).rejects.toMatchSnapshot();
    });
  });

  describe('getForecastForCountryOrCityInDateRange', () => {
    it('should return weather info for a country/city in data range', async () => {
      //given
      const locationNameInputDates: LocationNameInputDates = {
        locationType: LocationType.COUNTRY,
        locationName: 'country',
        from: '2023-02-15',
        to: '2023-02-17',
      };

      jest.spyOn(fetchDataApi, 'getGeoLocation').mockResolvedValue({
        latitude: 20,
        longitude: 20,
        location: LocationType.COUNTRY,
      });

      jest.spyOn(service, <any>'getDataApi').mockResolvedValue({
        timeArr: time,
        temperatureArr: temperature,
        weatherCodeArr: weathercode,
      });

      //when
      const result = await service.getForecastForCountryOrCityInDateRange(
        locationNameInputDates,
      );

      //then
      expect(result).toMatchSnapshot();
    });

    it('should throw error when unnamed country id dates range', async () => {
      //given
      const locationNameInputDates: LocationNameInputDates = {
        locationType: LocationType.COUNTRY,
        locationName: 'city',
        from: '2023-02-15',
        to: '2023-02-17',
      };

      jest.spyOn(fetchDataApi, 'getGeoLocation').mockImplementation(() => {
        throw new LocationException('Location not found');
      });

      await expect(
        service.getForecastForCountryOrCityInDateRange(locationNameInputDates),
      ).rejects.toMatchSnapshot();
    });
  });

  describe('getForecastForGeoLocationInDateRange', () => {
    it('should return weather response when valid coordinates', async () => {
      //given
      const coordinatesInputDates: CoordinatesInputDates = {
        lat: 20,
        lon: 20,
        from: '2023-02-15',
        to: '2023-02-17',
      };

      jest.spyOn(service, <any>'getDataApi').mockResolvedValue({
        timeArr: time,
        temperatureArr: temperature,
        weatherCodeArr: weathercode,
      });

      //when
      const result = await service.getForecastForGeoLocationInDateRange(
        coordinatesInputDates,
      );

      //then
      expect(result).toMatchSnapshot();
    });

    it('should throw error when problem with external api occur', async () => {
      //given
      const coordinatesInputDates: CoordinatesInputDates = {
        lat: 20,
        lon: 20,
        from: '2023-02-15',
        to: '2023-02-17',
      };

      jest.spyOn(service, <any>'getDataApi').mockImplementation(() => {
        throw new InternalServerException('Internal server error');
      });

      await expect(
        service.getForecastForGeoLocationInDateRange(coordinatesInputDates),
      ).rejects.toMatchSnapshot();
    });
  });
});
