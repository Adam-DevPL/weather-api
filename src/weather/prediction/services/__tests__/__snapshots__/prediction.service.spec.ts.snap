// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PredictionService getForecastForCountryOrCityInDateRange should return weather info for a country/city in data range 1`] = `
{
  "location": "country",
  "weatherInfo": [
    {
      "avgTemperature": "1.5",
      "day": "2023-02-15",
      "weather": "Fog and depositing rime fog",
    },
    {
      "avgTemperature": "2.3",
      "day": "2023-02-16",
      "weather": "Mainly clear, partly cloudy, and overcast",
    },
    {
      "avgTemperature": "5.3",
      "day": "2023-02-17",
      "weather": "Rain: Slight, moderate and heavy intensity",
    },
  ],
}
`;

exports[`PredictionService getForecastForCountryOrCityInDateRange should throw error when unnamed country id dates range 1`] = `[LocationException: Location not found]`;

exports[`PredictionService getForecastForGeoLocationInDateRange should return weather response when valid coordinates 1`] = `
{
  "location": {
    "latitude": 20,
    "longitude": 20,
  },
  "weatherInfo": [
    {
      "avgTemperature": "1.5",
      "day": "2023-02-15",
      "weather": "Fog and depositing rime fog",
    },
    {
      "avgTemperature": "2.3",
      "day": "2023-02-16",
      "weather": "Mainly clear, partly cloudy, and overcast",
    },
    {
      "avgTemperature": "5.3",
      "day": "2023-02-17",
      "weather": "Rain: Slight, moderate and heavy intensity",
    },
  ],
}
`;

exports[`PredictionService getForecastForGeoLocationInDateRange should throw error when problem with external api occur 1`] = `[InternalServerException: Internal server error]`;

exports[`PredictionService getForecastForSingleDay should return weather response for valid city - forecast on 1 day 1`] = `
{
  "location": "city",
  "weatherInfo": {
    "avgTemperature": "20.0",
    "day": "2023-02-15",
    "weather": "Mainly clear, partly cloudy, and overcast",
  },
}
`;

exports[`PredictionService getForecastForSingleDay should return weather response for valid country - forecast on 1 day 1`] = `
{
  "location": "country",
  "weatherInfo": {
    "avgTemperature": "20.0",
    "day": "2023-02-15",
    "weather": "Mainly clear, partly cloudy, and overcast",
  },
}
`;

exports[`PredictionService getForecastForSingleDay should throw error, when unnamed city - forecast on 1 day 1`] = `[LocationException: Location not found]`;

exports[`PredictionService getForecastForSingleDay should throw error, when wants country, but get city - forecast on 1 day 1`] = `[LocationTypeException: It is not a COUNTRY: city]`;
